### PR TITLE
Fix S6966 FP: FluentValidation ValidateAsync should not be proposed

### LINKED ISSUE #9339: Fix S6966 FP: FluentValidation ValidateAsync should not be proposed
Originally reported at https://community.sonarsource.com/t/c-s6966-with-fluentvalidation/116030.

### Description

Fluent validation supports asynchronous validation rules as described [here](https://docs.fluentvalidation.net/en/latest/async.html).
The validator fails with an exception if asynchronous rules are present and `Validate()` is called. In most cases, validation is synchronous, and sometimes, the context requires synchronous processing:
> Warning
If your validator contains asynchronous validators or asynchronous conditions, it’s important that you always call ValidateAsync on your validator and never Validate. If you call Validate, then an exception will be thrown.
You should not use asynchronous rules when [using automatic validation with ASP.NET](https://docs.fluentvalidation.net/en/latest/aspnet.html) as ASP.NET’s validation pipeline is not asynchronous. If you use asynchronous rules with ASP.NET’s automatic validation, they will always be run synchronously (10.x and older) or throw an exception (11.x and newer).

[Source](https://docs.fluentvalidation.net/en/latest/async.html)

FluentValidation has built-in guards about properly using "Validate vs. ValidateAsync". ValidateAsync is even the wrong choice in some circumstances, so S6966 has a high false positive ratio for this API.

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #9339

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Extensions/CompilationExtensions.cs [hunk 1]
```diff
@@ -23,7 +23,7 @@ namespace SonarAnalyzer.Extensions;
 internal static class CompilationExtensions
 {
     public static INamedTypeSymbol GetTypeByMetadataName(this Compilation compilation, KnownType knownType) =>
-        compilation.GetTypeByMetadataName(knownType.FullName);
+        compilation.GetTypeByMetadataName($"{knownType.FullName}{(knownType.GenericParameters.Any() ? $"`{knownType.GenericParameters.Count}" : string.Empty)}");
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/CompilationExtensions.cs:26]: Consider moving this logic to `KnownType.MetadataName`

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs [hunk 2]
```diff
@@ -31,6 +31,8 @@ public sealed partial class KnownType
         public string TypeName { get; }
         public string FullName { get; }
         public bool IsArray { get; init; }
+        public IReadOnlyList<string> GenericParameters => genericParameters;
+        public string MetadataName => $"{FullName}{(GenericParameters.Any() ? $"`{GenericParameters.Count}" : string.Empty)}";
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.Implementation.cs:35]: https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.isymbol.metadataname?view=roslyn-dotnet-4.6.0

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UseAwaitableMethod.cs [hunk 3]
```diff
@@ -86,7 +87,27 @@ private static WellKnownExtensionMethodContainer BuildWellKnownExtensionMethodCo
...
+    {
+        var exclusions = ImmutableArray.CreateBuilder<Func<IMethodSymbol, bool>>();
+        if (compilation.GetTypeByMetadataName(KnownType.Microsoft_EntityFrameworkCore_DbSet_TEntity) is not null)
+        {
+            exclusions.Add(x => x.IsAny(KnownType.Microsoft_EntityFrameworkCore_DbSet_TEntity, ExcludedMethodNamesAddAddRange)); // https://github.com/SonarSource/sonar-dotnet/issues/9269
+            exclusions.Add(x => x.IsAny(KnownType.Microsoft_EntityFrameworkCore_DbContext, ExcludedMethodNamesAddAddRange));     // https://github.com/SonarSource/sonar-dotnet/issues/9269
+        }
+        if (compilation.GetTypeByMetadataName(KnownType.FluentValidation_IValidator) is not null)
+        {
+            exclusions.Add(x => x.IsImplementingInterfaceMember(KnownType.FluentValidation_IValidator, "Validate"));   // https://github.com/SonarSource/sonar-dotnet/issues/9339
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/UseAwaitableMethod.cs:100]: `IsImplementingInterfaceMember` is a bit more expensive than a simple `Is` or `IsAny`. I, therefore, change the implementation to follow the same patterns as `BuildWellKnownExtensionMethodContainers`: create the lookup conditionally on compilation start and only add conditions that can potentially be true for the current compilation.

**Review by mary-georgiou-sonarsource — APPROVED** [review]
LGTM

**Comment by martin-strecker-sonarsource** [comment]
Peach validation: No changes as expected. We don't have projects with FluentValidation on peach.

**Pull request merged**